squint.c

```diff
@@ -0,0 +1,1611 @@
+}
+
+/**********************************************************/
+/********** peephole optimization funcs *******************/
+/**********************************************************/
+
+void apply_peepholes1(int *funcBegin, int *funcEnd)
+{
+   int *scan;
+   int *scanp1, *scanp2, *scanp3;
+   for (scan = funcBegin; scan < funcEnd; ++scan) {
+      scan = skip_nop(scan, 1);
+      /* one instruction peepholes */
+      if (*scan == 0xe2400000 || /* sub r0, r0, #0 */
+          *scan == 0xe2800000) { /* add r0, r0, #0 */
+         *scan = NOP;
+      }
+      else {    /* two instruction peephole */
+
+         scanp1 = active_inst(scan, 1);
+
+         if ((*scan & 0xfff00f00) == 0xe2800000 || // add rI, rS, #X
+             (*scan & 0xfff00f00) == 0xe2400000) { // sub rI, rS, #X
+            int rS = (*scan >> 16) & 0xf; // Source, RI_Rn
+            int rI = (*scan >> 12) & 0xf; // Index,  RI_Rd
+            if ((*scanp1 & 0xff3000ff) == 0xe5100000 &&
+                ((*scanp1 >> 16) & 0xf) == rI) { // ldr[b] rX, [rI]
+
+               *scanp1 = (*scanp1 & 0xff70ff00) | (*scan & 0xff) |
+                         (((*scan & 0xfff00000) == 0xe2800000) ? (1<<23) : 0)
+                         | (rS << 16); // ldr[b] rX, [rS, #X]
+               *scan = NOP;
+            }
+         }
+      }
+   }
+
+   funcEnd -= 2;
+   for (scan = funcBegin; scan < funcEnd; ++scan) {
+      scan = skip_nop(scan, 1);
+      scanp1 = active_inst(scan,   1);
+      scanp2 = active_inst(scanp1, 1);
+
+      /* change register to immediate if possible */
+      if (*scan   == 0xe52d0004 && /* push {r0} */
+          *scanp2 == 0xe49d1004 && /* pop  {r1} */
+          (*scanp1 & 0xffffff00) == 0xe3a00000) { /* mov r0, #X */
+         scanp3 = active_inst(scanp2, 1);
+         if ((*scanp3 & 0xfe1fffff) == 0xe0010000 &&
+             (*scanp3>>23 & 0x3) != 0x2) { /* exclude comparisons */
+            *scanp3 = (*scanp3 ^ 1<<16) | 1<<25 | (*scanp1 & 0xff);
+            *scan   = NOP;
+            *scanp1 = NOP;
+            *scanp2 = NOP;
+            scan = scanp3;
+         }
+         else if (*scanp3 == 0xe1a00051) { /* asr r0, r1, r0 */
+            int shift = ((*scanp1 & 0xff) < 0x20) ? (*scanp1 & 0x1f) : 0x1f;
+            *scanp3 = 0xe1a00040 | (shift << 7);
+            *scan   = NOP;
+            *scanp1 = NOP;
+            *scanp2 = NOP;
+            scan = scanp3;
+         }
+      }
+   }
+   funcEnd += 2;
+}
+
+void apply_peepholes2(int *instInfo, int *funcBegin, int *funcEnd)
+{
+   int *scan;
+   int *scanp1, *scanp2, *scanp3, *scanp4, *scanp5;
+   int *scanm1, *scanm2;
+
+   create_inst_info(instInfo, funcBegin, funcEnd);
+
+   funcEnd -= 6;
+   for (scan = funcBegin; scan < funcEnd; ++scan) {
+      scan = skip_nop(scan, 1);
+      /* Convert amacc array addressing to more compact form */
+      if ((*scan & 0xffffff00) == 0xe3a00000) { /* mov r0, #X */
+         int imm = *scan & 0xff;
+         scanm1 = active_inst(scan, -1);
```

Comment:
I think you should check whether you may underflow `funcBegin` when you move backward.

Also, as a general design question, I think subtracting `funcEnd` by 6 on Line 717 won't be safe enough.  It is possible that the function ends with NOP and then `active_inst(scan, 1)` will overflow beyond `funcEnd`.

Reply:
Every function begins with "push {fp, lr}; add fp, sp, #0" and ends with "add sp, fp, #0; pop {fp, pc}"  If you look carefully at the conditions being checked, I believe it is impossible to back up too far or go too far forward with the existing peephole functions, and given what code sequences AMaCC can currently generate legally.  For example, you will never find a "pop {r1}" as an instruction immediately before the ending "add sp, fp, #0; pop {fp, pc}" sequence.

Reply:
> given what code sequences AMaCC can currently generate legally

While this assumption may be true for now, it may be changed (or invalidated) when the code evolves in the future.  IMO, this kind of implicit assumption is difficult to keep track of.  I, personally, would rather start the loop after two instructions so that the reader can verify the correctness without checking the global context.

Also, I think the second half of my comment is not addressed.  If, for example, `apply_peepholes1` optimizes several instructions right before the return instruction, there will be more `NOP` instructions.  The `active_inst` on Line 730-732 risks out-of-bound array access because `funcEnd -= 6` above doesn't account for the number of NOP instructions in `(funcEnd-6, funcEnd]`.